src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+        \<down> M; (partial_case M of
+          Cont IntersectNotFound \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
+          Cont (IntersectFound _) \<Rightarrow>
+            client_program ItemsUpdate \<kappa> \<psi> \<I>
+        ) |
+      ItemsUpdate \<Rightarrow>
+        \<up> Cont RequestNext;
+        \<down> M; (partial_case M of
+          Cont (RollForward i) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (\<I> @ [i]) |
+          Cont (RollBackward p) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (roll_back_to \<I> p \<psi>) |
+          Cont AwaitReply \<Rightarrow> \<comment> \<open>client is up to date\<close>
+            \<up> Done;
+            \<bottom>
+        )
+    )"
+
+definition index_of :: "'p \<Rightarrow> 'i list \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> nat" where
+  [simp]: "index_of p \<I> \<psi> = (THE k. \<psi> (\<I> ! k) = p)"
+
+corec server_program where
+  "server_program rp mrb \<rho> \<psi> \<I> =
+    \<down> M; (partial_case M of
+      Done \<Rightarrow>
+        \<bottom> |
+      Cont (FindIntersect ps) \<Rightarrow> (
+        case \<rho> \<I> ps of
+          None \<Rightarrow>
+            \<up> Cont IntersectNotFound;
+            server_program rp mrb \<rho> \<psi> \<I> |
+          Some p \<Rightarrow>
+            \<up> Cont (IntersectFound p);
+            server_program (index_of p \<I> \<psi>) True \<rho> \<psi> \<I>
+      ) |
+      Cont RequestNext \<Rightarrow>
+        if mrb then
+          \<up> Cont (RollBackward (\<psi> (\<I> ! rp)));
+          server_program (Suc rp) False \<rho> \<psi> \<I>
+        else
+          if rp < length \<I> then
+            \<up> Cont (RollForward (\<I> ! rp));
+            server_program (Suc rp) mrb \<rho> \<psi> \<I>
+          else \<comment> \<open>client is up to date\<close>
+            \<up> Cont AwaitReply;
+            server_program rp mrb \<rho> \<psi> \<I>
+    )"
+
+context chain_sync
+begin
+
+primrec program where
+  "program Client = client_program IntersectionFinding candidate_points point_of initial_items\<^sub>c" |
+  "program Server = server_program 0 False best_intersection_point point_of initial_items\<^sub>s"
+
+end
+
+sublocale chain_sync \<subseteq> protocol_programs \<open>possibilities\<close> \<open>program\<close>
+proof
+  have "
+    client_program phase candidate_points point_of initial_items\<^sub>c
+    \<Colon>\<^bsub>Client\<^esub>
+    Cont possibilities" for phase
+    by
+      (coinduction arbitrary: initial_items\<^sub>c phase rule: up_to_embedding_is_sound)
+      (state_machine_bisimulation
+        program_expansion: client_program.code
+        extra_splits: or_done.splits message.splits phase.splits
+      )
+  moreover
+  have "
+    server_program read_ptr must_roll_back best_intersection_point point_of initial_items\<^sub>s
+    \<Colon>\<^bsub>Server\<^esub>
+    Cont possibilities" for read_ptr and must_roll_back
```

Comment:
The `for` clause should be put on a line of its own, as it refers to the whole preceding proposition and not just to the last bit of it that is on this line.

Reply:
Fixed in 06dec3e8d6c12f76eb26259a2544fe2848a0fe62, please check.